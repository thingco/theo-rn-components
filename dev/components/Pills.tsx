import React from "react";

import { Text } from "./Base";
import s, { FONT_SIZE_SMALL } from "./Styles";
import { TextStyle, StyleProp } from "react-native";

type PillType = "primary" | "contrast" | "warning";

type PillProps = {
  children: React.ReactNode;
  pillType: PillType;
  style?: StyleProp<TextStyle>;
};

const basePillStyles = [
  s.base_font_bold,
  s.text_sm,
  s.bold,
  s.overflow_hidden,
  s.pv_xs,
  s.ph_base,
  s.spacing_half,
  s.text_centre,
  { borderRadius: FONT_SIZE_SMALL },
];

function pillStyle(pillType: PillType): TextStyle[] {
  switch (pillType) {
    case "primary":
      return [...basePillStyles, s.white, s.bg_light];
    case "contrast":
      return [...basePillStyles, s.white, s.bg_contrast];
    case "warning":
      return [...basePillStyles, s.dark, s.bg_warning];
  }
}

const Pill: React.FC<PillProps> = ({
  pillType,
  children,
  ...props
}: PillProps) => (
  <Text style={[...pillStyle(pillType), props.style]}>{children}</Text>
);

function pillOutlineStyle(pillType: PillType): TextStyle[] {
  switch (pillType) {
    case "primary":
      return [...basePillStyles, s.border_2, s.light, s.border_light];
    case "contrast":
      return [...basePillStyles, s.border_2, s.contrast, s.border_contrast];
    case "warning":
      return [...basePillStyles, s.border_2, s.warning, s.border_warning];
  }
}

const PillOutlined: React.FC<PillProps> = ({
  pillType,
  children,
  ...props
}: PillProps) => (
  <Text style={[...pillOutlineStyle(pillType), props.style]}>{children}</Text>
);

function pillTintedStyle(pillType: PillType): TextStyle[] {
  switch (pillType) {
    case "primary":
      return [...basePillStyles, s.dark, s.bg_light_tinted];
    case "contrast":
      return [...basePillStyles, s.dark, s.bg_contrast_tinted];
    case "warning":
      return [...basePillStyles, s.dark, s.bg_warning_tinted];
  }
}

const PillTinted: React.FC<PillProps> = ({
  pillType,
  children,
  ...props
}: PillProps) => (
  <Text style={[...pillTintedStyle(pillType), props.style]}>{children}</Text>
);

export { Pill, PillOutlined, PillTinted };
