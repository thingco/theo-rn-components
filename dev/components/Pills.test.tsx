import React from "react";
import renderer from "react-test-renderer";

import { Pill, PillOutlined, PillTinted } from "./Pills";

test("Normal pill renders correctly", () => {
  const tree = renderer
    .create(<Pill pillType="primary">I'm a pill!</Pill>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Outline pill renders correctly", () => {
  const tree = renderer
    .create(<PillOutlined pillType="contrast">I'm a pill!</PillOutlined>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Tinted pill renders correctly", () => {
  const tree = renderer
    .create(<PillTinted pillType="warning">I'm a pill!</PillTinted>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
