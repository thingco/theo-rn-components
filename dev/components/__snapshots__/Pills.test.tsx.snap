// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normal pill renders correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Bariol",
      },
      Object {
        "fontSize": 16,
      },
      Array [
        Object {
          "fontFamily": "BariolBold",
        },
        Object {
          "fontSize": 14,
        },
        Object {
          "fontWeight": "bold",
        },
        Object {
          "overflow": "hidden",
        },
        Object {
          "paddingVertical": 4,
        },
        Object {
          "paddingHorizontal": 16,
        },
        Object {
          "letterSpacing": 0.5,
        },
        Object {
          "textAlign": "center",
        },
        Object {
          "borderRadius": 14,
        },
        Object {
          "color": "#FFFFFF",
        },
        Object {
          "backgroundColor": "#19988B",
        },
        undefined,
      ],
    ]
  }
>
  I'm a pill!
</Text>
`;

exports[`Outline pill renders correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Bariol",
      },
      Object {
        "fontSize": 16,
      },
      Array [
        Object {
          "fontFamily": "BariolBold",
        },
        Object {
          "fontSize": 14,
        },
        Object {
          "fontWeight": "bold",
        },
        Object {
          "overflow": "hidden",
        },
        Object {
          "paddingVertical": 4,
        },
        Object {
          "paddingHorizontal": 16,
        },
        Object {
          "letterSpacing": 0.5,
        },
        Object {
          "textAlign": "center",
        },
        Object {
          "borderRadius": 14,
        },
        Object {
          "borderWidth": 2,
        },
        Object {
          "color": "#FF6D37",
        },
        Object {
          "borderColor": "#FF6D37",
        },
        undefined,
      ],
    ]
  }
>
  I'm a pill!
</Text>
`;

exports[`Tinted pill renders correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Bariol",
      },
      Object {
        "fontSize": 16,
      },
      Array [
        Object {
          "fontFamily": "BariolBold",
        },
        Object {
          "fontSize": 14,
        },
        Object {
          "fontWeight": "bold",
        },
        Object {
          "overflow": "hidden",
        },
        Object {
          "paddingVertical": 4,
        },
        Object {
          "paddingHorizontal": 16,
        },
        Object {
          "letterSpacing": 0.5,
        },
        Object {
          "textAlign": "center",
        },
        Object {
          "borderRadius": 14,
        },
        Object {
          "color": "#003E52",
        },
        Object {
          "backgroundColor": "#f9e17f",
        },
        undefined,
      ],
    ]
  }
>
  I'm a pill!
</Text>
`;
